core/src/main/java/me/aurium/beetle/api/task/sync/SyncFuture.java

```diff
@@ -0,0 +1,72 @@
+package me.aurium.beetle.api.task.sync;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.function.Supplier;
+
+/**
+ * i want to stop existing
+ *
+ * A safer to use Future that permits get and join based on an internal behavior.
+ *
+ * TODO eventually some sort of resync to main thread functionality
+ *
+ * @param <T>
+ */
+public class SyncFuture<T> extends CompletableFuture<T> implements Synced {
+
+    private final Thread mainThread;
+    private final SyncQueue activeQueue;
+    private final BlockingBehavior behavior;
+
+    public SyncFuture(Thread mainThread, SyncQueue activeQueue, BlockingBehavior behavior) {
+        this.mainThread = mainThread;
+        this.activeQueue = activeQueue;
+        this.behavior = behavior;
+    }
+
+    //TODO a248 i don't know if this is necessary but given that get operates on a similar principle (wait on the main thread if we are running sync)
+    //i think it might be required. Anyways review this and see if it is necessary
+    @Override
+    public T get() throws ExecutionException, InterruptedException {
+        if (isOnMainThread()) {
+            while (!isDone()) {
+                activeQueue.pulseQueue();
+
+                if (isDone()) {
+                    return super.get();
+                }
+
+                behavior.behavior();
+
+            }
+        }
+
+        return super.get();
+    }
+
+    @Override
+    public T join() {
+        if (isOnMainThread()) {
+            while (!isDone()) {
+                activeQueue.pulseQueue(); //this is stepping in for the main sync-loop powered by the ticker timer
+                //when this is finished the ticker timer should return to ticking its timer and automatically handling the queue
+
+                if (isDone()) {
```

Comment:
There are some minor improvements you can make to avoid a double-volatile read between isDone() and join(). I've done so in [ManagedWaitStrategies](https://github.com/A248/ManagedWaitStrategies/tree/master/src/main/java/space/arim/managedwaits). But really, you are re-implementing a lot of things, which are not entirely simple to re-implement.

Reply:
Why would a double volatile read occur in between isDone and join?

Reply:
I'm not really attempting to re-implement anything, this is meant to be a simple and clean way to

A. learn more about threads
B. set up a future that can be joined on the main thread